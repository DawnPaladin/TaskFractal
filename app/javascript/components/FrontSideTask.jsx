import PropTypes from 'prop-types';
import React from 'react';
import * as Icon from 'react-feather';
import classNames from 'classnames';

import Checkbox from './Checkbox';
import deleteTask from './deleteTask';

// This component is placed either on a BackSideTask component, in an Outline, or in NextUp.
// The checked/unchecked state must be controlled by the parent.
export default class FrontSideTask extends React.Component {
	static propTypes = {
		task: PropTypes.object.isRequired,
		checkboxChange: PropTypes.func.isRequired,
		disableDescendantCount: PropTypes.bool,
	}
	constructor(props) {
		super(props);
		this.name = this.props.task.name; // for ease of debugging
		this.handleCheckbox = this.handleCheckbox.bind(this);
	}
	
	// Descendant count will come in as descendant_count or descendants.length, depending on where the component is included. Same for completed_descendant_count. Here we standardize these for easier access.
	getDescendantCount = () => {
		return Array.isArray(this.props.task.descendants) ? this.props.task.descendants.length : this.props.task.descendant_count
	}
	getCompletedDescendantCount = () => {
		return Array.isArray(this.props.task.completed_descendants) ? this.props.task.completed_descendants.length : this.props.task.completed_descendant_count;
	}
	
	handleCheckbox(event) {
		const task = {...this.props.task};
		task.completed = !task.completed;
		this.props.checkboxChange(task);
		// Parent component must call this.setState(), putting task wherever it belongs in parent's state.
		// Parent should also call send(task) as appropriate.
	}
	
	render() {
		var task = this.props.task;
		var url = "/tasks/" + task.id;
		var attachmentTitle = task.attachment_count > 1 ? task.attachment_count + " attachments" : "1 attachment";
		
		var underlineWidth, underline = null;
		if (this.getDescendantCount() > 0) {
			if (task.completed === true) underlineWidth = "100%";
			else if (this.getDescendantCount() === 0) underlineWidth = "0%"; 
			else underlineWidth = this.getCompletedDescendantCount()/this.getDescendantCount()*100+"%";
			underline = <div>
				<div className="task-underline-incomplete"></div>
				{/* <div className="task-underline-changing" style={{ width: underlineWidth }}></div> */}
				<div className="task-underline-complete" style={{ width: underlineWidth }}></div>
			</div>
		}
		
		return (
			<div className="task-card-front" ref={this.props.innerRef}>
				{/* <button className="delete-attachment-button" onClick={this.deleteTask}><Icon.Trash2 size="16" /></button> */}
				<Checkbox checked={task.completed} handleChange={this.handleCheckbox} />
				<a className={classNames('task-link', { "deemphasize": task.blocked_by_count > 0 })} href={url}>
					{ task.name }
					{ underline }
				</a>
				<div className="details">
					{ task.blocked_by_count ? 
						<div title={"Blocked by " + task.blocked_by_count}><Icon.PauseCircle size="17" /> {task.blocked_by_count} </div> : null
					}
					{ task.blocking_count ? 
						<div title={"Blocking " + task.blocking_count}><Icon.AlertCircle size="16" /> {task.blocking_count}</div> : null
					}
					{ task.due_date ?
						<div title={"Due " + task.due_date}><Icon.Calendar size="16" /> {task.dueDate}</div> : ""
					}
					{ task.description ?
						<div title={task.description}><Icon.AlignLeft size="16" /></div> : ""
					}
					{ task.attachment_count ?
						<div title={attachmentTitle}><Icon.Paperclip size="16" /> {task.attachment_count}</div> : ""
					}
					{ !this.props.disableDescendantCount && parseInt(task.descendants.length) > 0 ? 
						<div><Icon.CheckSquare size="16" /> {task.completed_descendants.length}/{task.descendants.length}</div> : ""
					}
				</div>
			</div>
		)
	}
}
