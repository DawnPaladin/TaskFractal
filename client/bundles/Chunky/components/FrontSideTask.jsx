import PropTypes from 'prop-types';
import React from 'react';
import * as Icon from 'react-feather';

import Checkbox from './Checkbox';
import send from './send';
import deleteTask from './deleteTask';

// This component is placed either on a BackSideTask component or in an Outline.
export default class FrontSideTask extends React.Component {
	static propTypes = {
		task: PropTypes.object.isRequired,
		handleCheckboxChange: PropTypes.func,
		disableDescendantCount: PropTypes.bool,
	}
	constructor(props) {
		super(props);
		this.name = this.props.task.name; // for ease of debugging
		this.state = {
			task: this.props.task
		}
		this.handleCheckbox = this.handleCheckbox.bind(this);
		this.checkboxChange = this.props.handleCheckboxChange ? this.props.handleCheckboxChange.bind(this) : this.checkboxChange.bind(this);
		this.deleteTask = deleteTask.bind(this);
	}
	
	handleCheckbox(event) {
		this.checkboxChange(event, this);
	}
	
	// If FrontSideTask is in a BackSideTask, BackSideTask will provide a checkboxChange function. If not, we use this one.
	checkboxChange(event) {
		const completed = event.target.checked;
		this.setState(
			(prevState, props) => ({
				task: {
					...prevState.task,
					completed: completed
				}
			}),
			() => { send(this.state.task); }
		);
	}
		
	render() {
		var url = "/tasks/" + this.state.task.id;
		return (
			<div className="task-card-front" ref={this.props.innerRef}>
				{/* <button className="delete-attachment-button" onClick={this.deleteTask}><Icon.Trash2 size="16" /></button> */}
				<Checkbox checked={this.state.task.completed} handleChange={this.handleCheckbox} />
				<a className="task-link" href={url}>{ this.state.task.name }</a>
				<div className="details">
					{ this.state.task.due_date ?
						<div><Icon.Calendar size="16" /> {this.state.task.dueDate}</div> : ""
					}
					{ this.state.task.description ?
						<div><Icon.AlignLeft size="16" /></div> : ""
					}
					{ this.state.task.attachment_count ?
						<div><Icon.Paperclip size="16" /> {this.state.task.attachment_count}</div> : ""
					}
					{ !this.props.disableDescendantCount && parseInt(this.state.task.descendants.length) > 0 ? 
						<div><Icon.CheckSquare size="16" /> {this.state.task.completed_descendants.length}/{this.state.task.descendants.length}</div> : ""
					}
				</div>
			</div>
		)
	}
}
