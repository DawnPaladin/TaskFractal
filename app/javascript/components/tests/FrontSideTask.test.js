import React from 'react'; // JSX

import renderer from 'react-test-renderer'; // for snapshots

// for running tests on jQuery-like tests
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import FrontSideTask from '../FrontSideTask';

const task = {
	completed: true,
	name: "Test task",
	id: 0,
	attachment_count: 0,
	descendants: [
		{ name: "First descendant", id: 1, completed: false },
		{ name: "Second descendant", id: 2, completed: false }
	]
};
const checkboxChange = () => {};

jest.mock('../Checkbox'); // checkbox is from @material library which doesn't test well

test('FrontSideTask snapshot', () => {
	const component = renderer.create(
		<FrontSideTask task={task} checkboxChange={checkboxChange} disableDescendantCount={false} />
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('has descendants by default', () => {
	const wrapper = shallow( <FrontSideTask task={task} checkboxChange={checkboxChange} /> );
	const descendants = wrapper.find('.descendants');
	expect(descendants.exists()).toBeTruthy();
});

test('disableDescendantCount hides descendant count', () => {
	const wrapper = shallow( <FrontSideTask task={task} checkboxChange={checkboxChange} disableDescendantCount={true} /> );
	const descendants = wrapper.find('.descendants')
	expect(descendants.exists()).toBeFalsy();
});
