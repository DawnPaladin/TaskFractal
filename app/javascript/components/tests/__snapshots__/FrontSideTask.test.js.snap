// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontSideTask snapshot 1`] = `
<div
  className="task-card-front"
>
  <a
    className="task-link"
    href="/tasks/0"
  >
    Test task
    <div>
      <div
        className="task-underline-incomplete"
      />
      <div
        className="task-underline-complete"
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </div>
  </a>
  <div
    className="details"
  >
    
    
    
    <div
      className="descendants"
    >
      <svg
        fill="none"
        height="16"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 11 12 14 22 4"
        />
        <path
          d="M21 12v7a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11"
        />
      </svg>
       
      /
      2
    </div>
  </div>
</div>
`;
